### ISSUE TITLE
Game instance still alive even when widget has been disposed

### ISSUE THREAD

**Issue by khoafornia** [flame-engine/flame#834]
# Current bug behaviour

When navigating away from a screen displaying a Flame game widget, the instance remains alive and still runs
# Expected behaviour

When navigating away from a screen displaying a Flame game widget, the instance is disposed
# Steps to reproduce

1. Add a print statement inside of the BaseGame class's update method.
2. Route to a new page inside of the update method.
3. The print statement still prints and the screen is rerouted in a loop.

# Flutter doctor output

```
Output of: flutter doctor -v
```
[✓] Flutter (Channel stable, 2.2.0, on macOS 11.4 20F5046g darwin-arm, locale en-US)
    • Flutter version 2.2.0 at /Users/name/Developer/flutter
    • Framework revision b22742018b (2 weeks ago), 2021-05-14 19:12:57 -0700
    • Engine revision a9d88a4d18
    • Dart version 2.13.0

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/name/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

# More environment information

- Flame version: 1.0.0-releasecandidate.11

# Log information

```
Enter log information in this code block
```

# More information

- Possibly same bug: https://stackoverflow.com/questions/66704816/kill-every-flames-game-processes

**Comment by spydon** [comment]
Do you create the game instance outside of the widget?

**Comment by renancaraujo** [comment]
Flame games are like flutetrs controllers for example: Whoever instantiates should dispose of it.

**Comment by khoafornia** [comment]
I instantiated the game instance inside of the widget but oddly, Flutter is disposing the containing widget but the Flame instance isn't disposed. My workaround is commented out.

`class GameWrapper extends StatefulWidget {
  const GameWrapper({required Key key}) : super(key: key);

  @override
  _GameWrapperState createState() => _GameWrapperState();
}

class _GameWrapperState extends State<GameWrapper> {
  // MyGame? _newGameInstance = MyGame();
  //
  // @override
  // dispose() {
  //   _newGameInstance!.detach();
  //   _newGameInstance = null;
  //   super.dispose();
  // }

  @override
  Widget build(BuildContext context) {
    return GameWidget<MyGame>(
      game: MyGame(),
      //game: _newGameInstance(),
    );
  }
}`

**Comment by erickzanardo** [comment]
try calling the `pauseEngine` method on your state dispose method

**Comment by khoafornia** [comment]
Are we sure `pauseEngine` disposes the game instance? It seems to be a significant performance increase when `detach` is called instead of `pauseEngine`.

**Comment by erickzanardo** [comment]
`pauseEngine` will not disposes the instance, what it will do is to remove the callbacks scheduled on the rendering ticker.

Calling detach will also clear any allocated resources as we can see here: https://github.com/flame-engine/flame/blob/main/packages/flame/lib/src/game/game.dart#L119

The instance will live as long as your state instance lives.

tbh, I think what is missing on flame, is that we should call `pauseEngine` on the `detach` method, it makes sense to remove the scheduled ticker from a game instance that isn't attached anymore. @flame-engine/flame-colab wdyt?

**Comment by spydon** [comment]
> tbh, I think what is missing on flame, is that we should call `pauseEngine` on the `detach` method, it makes sense to remove the scheduled ticker from a game instance that isn't attached anymore. @flame-engine/flame-colab wdyt?
Makes sense!

**Comment by wenxiangjiang** [comment]
Is `dispose` method of `GameWrapper` invoked when navigating away from a screen?

**Comment by erickzanardo** [comment]
@khoafornia can you check #856 please? I've made a small widget test that would fail if the game wouldn't be correctly disposed after a navigation.

If I missed something and you are still facing this issue, could you add a minimal reproducible code where this issue still shows up?

**Comment by khoafornia** [comment]
@erickzanardo Verified! 😄
